### PR TITLE
Fix issue 2966

### LINKED ISSUE #2966: [Data Layer] Re-work `SyncExchangeRatesAct` to repository and use-case
### Please confirm the following

- [X] I checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.

### What would you like to improve?

- This legacy class https://github.com/Ivy-Apps/ivy-wallet/blob/main/temp/legacy-code/src/main/java/com/ivy/legacy/domain/action/exchange/SyncExchangeRatesAct.kt refactored to UseCase, Repository, DataSource
- The above covered with unit tests 
- The above covered wirh integration tests (androidTest)
- The above being type-safe and using Arrow Typed-Erroea

### Because...

- getting rid of legacy code
- safety, stability and maintainability
- migration to our new architecture

### Description

1. Check https://github.com/Ivy-Apps/ivy-wallet/blob/main/temp/legacy-code/src/main/java/com/ivy/legacy/domain/action/exchange/SyncExchangeRatesAct.kt
2. In `:shared-data` create `ExchangeRatesRepository` that:
- Fetches exchange rates from a `RemoteExchangeRatesDataSource` (the Ktor request)
- Can save exchange rates locally in room using directly the daos that we already have
- Can save exchange rates overrides
- Can read exchange rates
- It should map DB exchange rates entity to a new domain model that is - type-safe and explicitly defined! (see our `data.model` package)
3. In `:shared-domain` create `SyncExchangeRatesUseCase` without inherenting anything and simply adding some `suspend fun sync()` method (does the same as https://github.com/Ivy-Apps/ivy-wallet/blob/main/temp/legacy-code/src/main/java/com/ivy/legacy/domain/action/exchange/SyncExchangeRatesAct.kt)
4. Cover the above with unit tests
5. Cover the use-case with integration tests
6. Migrate https://github.com/Ivy-Apps/ivy-wallet/blob/main/temp/legacy-code/src/main/java/com/ivy/legacy/domain/action/exchange/SyncExchangeRatesAct.kt usages and delete it.

### Success Criteria

- Read https://arrow-kt.io/learn/typed-errors/working-with-typed-errors/ and make sure to work with typed-errors. **Do NOT throw exceptions, use Either from Arrow**
- Cover all classes written with unit tests using given-when-then styling (see existing tests)
- Cover the use case with an integration test (see how androidTest is setup in `:shared:data` + you must update the Integration Test GitHub Action workflow because currently it runs integration tests only for `:shared:data`)
- The legacy code about exchange rates is deleted and the app works as before but using `SyncExchangeRatesUseCase` and the `ExchangeRatesRepository`

### PR DESCRIPTION
**Pull request by AMHaroun**
## Pull Request (PR) Checklist
Please check if your pull request fulfills the following requirements:
- [x] The PR is submitted to the `main` branch.
- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read the [Architecture Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Architecture.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] 🎬 I've attached a **screen recoding** of the changes. 

https://imgur.com/a/4ngl9Y0

> Tip: drag & drop the video to the PR description.

## What's changed?

Describe with a few bullets **what's new:**
- a Refactored SyncExchangeRatesAct
- b Wrote Unit & Integration Tests
- c Removed Legacy code & replaced with new Use Case, Repository, & Data source

> 💡 Tip: Please, attach screenshots and screen recordings. It helps a lot!

## Risk Factors

**What may go wrong if we merge your PR?**

- a
- b

**In what cases your code won't work?**

- a
- b

## Does this PR closes any GitHub Issues?

Check **[Ivy Wallet Issues](https://github.com/Ivy-Apps/ivy-wallet/issues)**.

- Closes #2966 

> Replace `ISSUE_NUMBER` with the number of the GitHub issue that this PR is fixing. If you've done that correctly, you'll see the issue title linked when previewing your PR description.

> **!Note:** Do **not** link the PR number. Link the number/id of the GitHub Issue from [issues](https://github.com/Ivy-Apps/ivy-wallet/issues).

## Troubleshooting CI failures ❌

GitHub Actions failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

**Comment by ILIYANGERMANOV** [comment]
Hey @AMHaroun you need to resolve merge conflicts and I'll review you after the CI pass

File: shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import dagger.hilt.components.SingletonComponent
+import io.ktor.client.HttpClient
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object RemoteDataSourceModule {
+    @Provides
+    @Singleton
+    fun provideRemoteExchangeRatesDataSource(ktorClient: HttpClient): RemoteExchangeRatesDataSource {
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt:17]: No neee for that, you can use `dagger.Lazy<HttptClient>`

File: shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.ivy.data.model
+
+data class ExchangeRate(
+    val baseCurrency: String,
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt:4]: use `AssetCode`

File: shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    val currency: String,
+    val rate: Double,
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt:6]: use `PositiveDouble`

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import arrow.core.Either
+import com.ivy.data.remote.RemoteExchangeRatesDataSource
+import com.ivy.data.remote.responses.ExchangeRatesResponse
+import io.ktor.client.HttpClient
+import io.ktor.client.call.body
+import io.ktor.client.request.get
+import javax.inject.Inject
+
+class RemoteExchangeRatesDataSourceImpl
+    @Inject
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:12]: formatting: don't put it on a separate line cuz it looks a bit harder to read

File: shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    constructor(
+        private val ktorClient: Lazy<HttpClient>,
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/remote/impl/RemoteExchangeRatesDataSourceImpl.kt:14]: Use `dagger.Lazy` and remove the `@Provides`

File: shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.ivy.data.remote.responses
+
+data class ExchangeRatesResponse(
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/remote/responses/ExchangeRatesResponse.kt:3]: This should be `@Serializable` and also have `@Keep`

File: shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import com.ivy.data.model.ExchangeRate
+import com.ivy.data.remote.responses.ExchangeRatesResponse
+import kotlinx.coroutines.flow.Flow
+
+interface ExchangeRatesRepository {
+    val urls: List<String>
+
+    suspend fun fetchExchangeRates(): ExchangeRatesResponse?
+
+    suspend fun save(value: ExchangeRateEntity)
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt:13]: The repository should work only with domain objects, don't expose entities

File: shared/data/src/main/java/com/ivy/data/repository/impl/ExchangeRatesRepositoryImpl.kt [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+class ExchangeRatesRepositoryImpl
+    @Inject
+    constructor(
+        private val mapper: ExchangeRateMapper,
+        private val exchangeRatesDao: ExchangeRatesDao,
+        private val writeExchangeRatesDao: WriteExchangeRatesDao,
+        private val remoteExchangeRatesDataSource: RemoteExchangeRatesDataSource,
+        private val dispatchersProvider: DispatchersProvider,
+    ) : ExchangeRatesRepository {
+        override val urls =
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/impl/ExchangeRatesRepositoryImpl.kt:26]: This urla logic doesn't belong to the repository. It's a responsibility of the `RemoteExchangeRatesDataSource` so you should encapsulate it there

File: shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.ivy.data.db.entity.ExchangeRateEntity
+import com.ivy.data.model.ExchangeRate
+import com.ivy.data.remote.responses.ExchangeRatesResponse
+import kotlinx.coroutines.flow.Flow
+
+interface ExchangeRatesRepository {
+    val urls: List<String>
+
+    suspend fun fetchExchangeRates(): ExchangeRatesResponse?
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt:11]: Do not expose this response. The repository should work only with domain models. Create a domain model using our exact types (e.g. `AssetCode`, `PositiveDouble`) and map dtos -> domaij models

File: shared/data/src/main/java/com/ivy/data/repository/mapper/ExchangeRateMapper.kt [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package com.ivy.data.repository.mapper
+
+import com.ivy.data.db.entity.ExchangeRateEntity
+import com.ivy.data.model.ExchangeRate
+import javax.inject.Inject
+
+class ExchangeRateMapper
+    @Inject
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/repository/mapper/ExchangeRateMapper.kt:8]: why on new line? put it on top like the other classes

it should be
```kotlin
class A @inject constructor(
   ...
) {
 ...
}
```

File: shared/domain/build.gradle.kts [hunk 11]
```diff
@@ -1,14 +1,23 @@
...
     implementation(projects.shared.data)
 
-    implementation(libs.datastore) // TODO: Will be removed
+    implementation(libs.datastore)
+    implementation(libs.bundles.ktor)
+
+    val mockkVersion = "1.13.10"
+    androidTestImplementation(libs.bundles.integration.testing)
+    androidTestImplementation(projects.shared.testing)
+    androidTestImplementation("io.mockk:mockk-android:$mockkVersion")
```
Comment (ILIYANGERMANOV) [inline shared/domain/build.gradle.kts:21]: Use the version catalog - all dependencies must be defined there - `libs.versions.toml`

**Review by ILIYANGERMANOV — COMMENTED** [review]
Excellent work! 💯 I left some comments that needs to be addressed before merging

**Comment by AMHaroun** [comment]
Okay on it 👍

File: shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+import io.ktor.client.HttpClient
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object RemoteDataSourceModule {
+    @Provides
+    @Singleton
+    fun provideRemoteExchangeRatesDataSource(ktorClient: dagger.Lazy<HttpClient>): RemoteExchangeRatesDataSource {
+        return RemoteExchangeRatesDataSourceImpl(ktorClient)
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt:18]: Why this provides? Use `@Binds` + `@Inject constructor`

Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt:18]: See how the other dataources or repositories are implemented:
- none instances are created manually by calling the constructor
- Hilt `@Binds`
- `@Inject constructor`
- this way if I add more params to the impl.class constructor it'll work out of the box and I won't have to modify some provide methods

File: shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.ivy.data.model
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.PositiveDouble
+
+data class ExchangeRate(
+    val baseCurrency: AssetCode,
+    val currency: String,
```
Comment (ILIYANGERMANOV) [inline shared/data/src/main/java/com/ivy/data/model/ExchangeRate.kt:8]: same here, this currency should be `AssetCode` it's not any arbitrary String

File: shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt (hunk 12 above)
Comment (AMHaroun) [reply shared/data/src/main/java/com/ivy/data/di/RemoteDataSourceModule.kt:18]: Oh okay I get it now

File: shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt (hunk 9 above)
Comment (ILIYANGERMANOV) [reply shared/data/src/main/java/com/ivy/data/repository/ExchangeRatesRepository.kt:9]: The `ExchangeRatesResponse` should not be exposed but we can fix that in a follow-up PR. This response is Dto and the repo should return only domain model

**Review by ILIYANGERMANOV — APPROVED** [review]
Can you please fix the unresolved comments in a follow-up PR? We can merge this one cuz it's great improvement

**Comment by AMHaroun** [comment]
> Can you please fix the unresolved comments in a follow-up PR? We can merge this one cuz it's great improvement

Yes sure, I will just try to fix CI quickly

**Comment by ILIYANGERMANOV** [comment]
@AMHaroun also "Update branch" to verify that the CI integrates well with the existing code

**Pull request merged**